### PR TITLE
Multifeatures

### PR DESCRIPTION
**Pull request by ghoshkaj**
https://github.com/mapbox/mapbox-gl-draw/issues/355

### CODE REVIEW

File: src/modes/simple_select.js [hunk 1]
```diff
@@ -233,16 +234,82 @@ module.exports = function(ctx, options = {}) {
...
+    combineFeatures: function() {
+      var selectedFeatures = ctx.store.getSelected();
+
+      if(!selectedFeatures || selectedFeatures.length < 2) return;
+
+      var featureType = selectedFeatures[0].type.replace('Multi','');
+      var coordinates = [];
+      var properties = selectedFeatures[0].properties;
+
+      selectedFeatures = selectedFeatures.map(function(feature) {
```
Comment (mcwhittemore) [inline src/modes/simple_select.js]: To be clear, this map is doing two things.

1. It is creating the coordinates array for the new MultiFeature
2. Getting the GeoJSON of the selected features that have the same non-multi type of the first feature

File: src/modes/simple_select.js [hunk 2]
```diff
@@ -233,16 +234,82 @@ module.exports = function(ctx, options = {}) {
...
+      var multiFeature = new MultiFeature(ctx, {
+        type: Constants.geojsonTypes.FEATURE,
+        properties: selectedFeatures[0].properties || {},
+        geometry: {
+          type: 'Multi' + featureType,
+          coordinates: coordinates
+        }
+      });
+      ctx.store.add(multiFeature);
+      ctx.store.delete(ctx.store.getSelectedIds(), { silent: true });
```
Comment (mcwhittemore) [inline src/modes/simple_select.js]: This is going to delete all of the selected features.

File: src/modes/simple_select.js [hunk 3]
```diff
@@ -233,16 +234,82 @@ module.exports = function(ctx, options = {}) {
...
+      ctx.store.setSelected([multiFeature.id]);
+
+      ctx.map.fire(Constants.events.MERGE, {
+        createdFeatures: [multiFeature.toGeoJSON()],
+        deletedFeatures: selectedFeatures
```
Comment (mcwhittemore) [inline src/modes/simple_select.js]: This is going to broadcast that the features deleted are only those that match the non-multi type of the first feature in the list.

File: src/modes/simple_select.js [hunk 4]
```diff
@@ -233,16 +234,82 @@ module.exports = function(ctx, options = {}) {
...
       push(geojson);
       if (geojson.properties.active !== Constants.activeStates.ACTIVE
         || geojson.geometry.type === Constants.geojsonTypes.POINT) return;
       createSupplementaryPoints(geojson).forEach(push);
     },
-    trash() {
+    trash: function() {
       ctx.store.delete(ctx.store.getSelectedIds());
+    },
+    combineFeatures: function() {
```
Comment (mcwhittemore) [inline src/modes/simple_select.js]: this needs a check that all of the selected features are of the same type. If they are not all of the same type, nothing should happen.

Comment (mcwhittemore) [inline src/modes/simple_select.js]: @samanpwbb do you agree a noop is right if the types don't match?

File: src/modes/simple_select.js [hunk 5]
```diff
@@ -233,16 +234,82 @@ module.exports = function(ctx, options = {}) {
...
+          });
+          ctx.store.delete(feature.id, { silent: true });
+        }
+
+        return feature.toGeoJSON();
+      })
+
+      ctx.map.fire(Constants.events.UNMERGE, {
+        createdFeatures: createdFeatures,
+        deletedFeatures: selectedFeatures
```
Comment (mcwhittemore) [inline src/modes/simple_select.js]: This is saying all features selected were deleted. Line 303 says that only features that are a MultiFeature get deleted.

File: src/modes/simple_select.js [hunk 6]
```diff
@@ -233,16 +234,82 @@ module.exports = function(ctx, options = {}) {
...
       ctx.store.delete(ctx.store.getSelectedIds());
+    },
+    combineFeatures: function() {
+      var selectedFeatures = ctx.store.getSelected();
+
+      if(!selectedFeatures || selectedFeatures.length < 2) return;
+
+      var featureType = selectedFeatures[0].type.replace('Multi','');
+      var coordinates = [];
+      var properties = selectedFeatures[0].properties;
```
Comment (mcwhittemore) [inline src/modes/simple_select.js]: This is not used.

File: src/modes/simple_select.js (hunk 4 above)
Comment (samanpwbb) [inline src/modes/simple_select.js]: Yeah, we should noop internally. Studio should dim the button and show "cannot merge features with different geometry types" in a popover.

File: src/ui.js [hunk 7]
```diff
@@ -153,6 +154,28 @@ module.exports = function(ctx) {
         }
       });
     }
+
+    if (controls.merge_features) {
+      buttonElements.trash = createControlButton('combineFeatures', {
```
Comment (mcwhittemore) [inline src/ui.js]: `buttonElements.trash` should not be `trash`

File: src/ui.js [hunk 8]
```diff
@@ -153,6 +154,28 @@ module.exports = function(ctx) {
...
+        className: Constants.classes.CONTROL_BUTTON_MERGE_FEATURES,
+        title: 'Merge',
+        onActivate: () => {
+          ctx.events.combineFeatures();
+        }
+      });
+    }
+
+    if (controls.split_features) {
+      buttonElements.trash = createControlButton('uncombineFeatures', {
```
Comment (mcwhittemore) [inline src/ui.js]: `buttonElements.trash` should not be `trash`

File: src/ui.js [hunk 9]
```diff
@@ -153,6 +154,28 @@ module.exports = function(ctx) {
         }
       });
     }
+
+    if (controls.merge_features) {
+      buttonElements.trash = createControlButton('combineFeatures', {
+        container: controlGroup,
+        className: Constants.classes.CONTROL_BUTTON_MERGE_FEATURES,
+        title: 'Merge',
```
Comment (mcwhittemore) [inline src/ui.js]: Combine

File: src/ui.js [hunk 10]
```diff
@@ -153,6 +154,28 @@ module.exports = function(ctx) {
...
+          ctx.events.combineFeatures();
+        }
+      });
+    }
+
+    if (controls.split_features) {
+      buttonElements.trash = createControlButton('uncombineFeatures', {
+        container: controlGroup,
+        className: Constants.classes.CONTROL_BUTTON_SPLIT_FEATURES,
+        title: 'Split',
```
Comment (mcwhittemore) [inline src/ui.js]: Uncombine

File: src/ui.js [hunk 11]
```diff
@@ -153,6 +154,28 @@ module.exports = function(ctx) {
...
+        onActivate: () => {
+          ctx.events.combineFeatures();
+        }
+      });
+    }
+
+    if (controls.split_features) {
+      buttonElements.trash = createControlButton('uncombineFeatures', {
+        container: controlGroup,
+        className: Constants.classes.CONTROL_BUTTON_SPLIT_FEATURES,
```
Comment (mcwhittemore) [inline src/ui.js]: Uncombine

File: src/ui.js [hunk 12]
```diff
@@ -153,6 +154,28 @@ module.exports = function(ctx) {
         }
       });
     }
+
+    if (controls.merge_features) {
+      buttonElements.trash = createControlButton('combineFeatures', {
+        container: controlGroup,
+        className: Constants.classes.CONTROL_BUTTON_MERGE_FEATURES,
```
Comment (mcwhittemore) [inline src/ui.js]: Combine

File: test/mode_handler.test.js [hunk 13]
```diff
@@ -158,3 +160,55 @@ test('Modehandler#trash without a mode.trash', t => {
 
   t.end();
 });
+
+// test('Modehandler#combineFeatures', t => {
```
Comment (mcwhittemore) [inline test/mode_handler.test.js]: We should drop these commented out tests

File: src/ui.js (hunk 12 above)
Comment (mcwhittemore) [inline src/ui.js]: Also, this class isn't in the constants file

File: src/ui.js (hunk 11 above)
Comment (mcwhittemore) [inline src/ui.js]: Also, this class isn't in the constants file

File: test/api.test.js [hunk 14]
```diff
@@ -422,6 +422,72 @@ test('Draw.modes', t => {
...
 
+test('Draw.combineFeatures -- multiploygon', t => {
+  const [polygonId] = Draw.add(getGeoJSON('polygon'));
+  const [polygon2Id] = Draw.add(getGeoJSON('polygon2'));
+  Draw.changeMode('simple_select', { featureIds: [polygonId, polygon2Id]});
+
+  Draw.combineFeatures();
+  t.equals(Draw.getAll().features.length, 1, 'can merge two features');
+  t.equals(Draw.getAll().features[0].geometry.type, 'MultiPolygon', 'can merge two polygons into MultiPolygon');
+  Draw.deleteAll();
```
Comment (mcwhittemore) [inline test/api.test.js]: Should we test the geometry?

File: test/api.test.js [hunk 15]
```diff
@@ -422,6 +422,72 @@ test('Draw.modes', t => {
...
+
+test('Draw.combineFeatures -- multipoint', t => {
+  const [pointId] = Draw.add(getGeoJSON('point'));
+  const [point2Id] = Draw.add(getGeoJSON('point2'));
+  Draw.changeMode('simple_select', { featureIds: [pointId, point2Id]});
+
+  Draw.combineFeatures();
+  t.equals(Draw.getAll().features.length, 1, 'can merge two features');
+  t.equals(Draw.getAll().features[0].geometry.type, 'MultiPoint', 'can merge two points into MultiPoint');
+  Draw.deleteAll();
```
Comment (mcwhittemore) [inline test/api.test.js]: Should we test the geometry?

File: test/api.test.js [hunk 16]
```diff
@@ -422,6 +422,72 @@ test('Draw.modes', t => {
...
+
+test('Draw.combineFeatures -- multilinestring', t => {
+  const [lineId] = Draw.add(getGeoJSON('line'));
+  const [line2Id] = Draw.add(getGeoJSON('line2'));
+  Draw.changeMode('simple_select', { featureIds: [lineId, line2Id]});
+
+  Draw.combineFeatures();
+  t.equals(Draw.getAll().features.length, 1, 'can merge two features');
+  t.equals(Draw.getAll().features[0].geometry.type, 'MultiLineString', 'can merge two linestrings into MultiLineString');
+  Draw.deleteAll();
```
Comment (mcwhittemore) [inline test/api.test.js]: Should we test the geometry?

File: test/api.test.js [hunk 17]
```diff
@@ -422,6 +422,72 @@ test('Draw.modes', t => {
...
+  t.end();
+});
+
+test('Draw.uncombineFeatures -- multilinestring', t => {
+  const [multiLineStringId] = Draw.add(getGeoJSON('multiLineString'));
+  Draw.changeMode('simple_select', { featureIds: [multiLineStringId]});
+
+  Draw.uncombineFeatures();
+  t.equals(Draw.getAll().features.length, 2, 'can split multiLineString');
+  Draw.deleteAll();
```
Comment (mcwhittemore) [inline test/api.test.js]: Should we test the geometry?

File: test/api.test.js [hunk 18]
```diff
@@ -422,6 +422,72 @@ test('Draw.modes', t => {
...
+  t.end();
+});
+
+test('Draw.uncombineFeatures -- multipolygon', t => {
+  const [multipolygon2Id] = Draw.add(getGeoJSON('multiPolygon2'));
+  Draw.changeMode('simple_select', { featureIds: [multipolygon2Id]});
+
+  Draw.uncombineFeatures();
+  t.equals(Draw.getAll().features.length, 2, 'can split multipolygon');
+  Draw.deleteAll();
```
Comment (mcwhittemore) [inline test/api.test.js]: Should we test the geometry?

File: test/api.test.js [hunk 19]
```diff
@@ -422,6 +422,72 @@ test('Draw.modes', t => {
...
+  t.end();
+});
+
+test('Draw.uncombineFeatures -- multipoint', t => {
+  const [multipointId] = Draw.add(getGeoJSON('multiPoint'));
+  Draw.changeMode('simple_select', { featureIds: [multipointId]});
+
+  Draw.uncombineFeatures();
+  t.equals(Draw.getAll().features.length, 2, 'can split multipoint');
+  Draw.deleteAll();
```
Comment (mcwhittemore) [inline test/api.test.js]: Should we test the geometry?

File: test/api.test.js (hunk 14 above)
Comment (ghoshkaj) [inline test/api.test.js]: What do you mean by this?

File: API.md [hunk 20]
```diff
@@ -244,16 +244,48 @@ var ids = Draw.set({type: 'FeatureCollection', features: [{
...
+This invokes the current modes trash event. For the `simple_select` mode this deletes all active features. For the `direct_select` mode this deletes the active vertices. For the drawing modes, this cancels the current process.
 
-This is different from `delete` or `deleteAlll` in that it follows rules described by the current mode.
+This is different from `delete` or `deleteAll` in that it follows rules described by the current mode.
+
+---
+
+### `.combineFeatures() -> Draw`
+
+This invokes the current modes combineFeatures event. For the `simple_select` mode this function will combine all single features into a multifeature, so long as they are all of the same geometry type. For example:
```
Comment (mcwhittemore) [inline API.md]: `will combine all single features`

We should not `selected` and is `single` correct?

Comment (mcwhittemore) [inline API.md]: Ah, single is a bit confusing. My question was cleared up by the next line, so lets drop `single`.

File: API.md [hunk 21]
```diff
@@ -244,16 +244,48 @@ var ids = Draw.set({type: 'FeatureCollection', features: [{
...
+
+- LineString, LineString => MultiLineString
+- MultiLineString, LineString => MultiLineString
+- MultiLineString, MultiLineString => MultiLineString
+
+Calling this function on different geometry types will not cause any changes. For example:
+
+- Point + LineString => no action taken
```
Comment (mcwhittemore) [inline API.md]: You used `,` above.

File: API.md [hunk 22]
```diff
@@ -244,16 +244,48 @@ var ids = Draw.set({type: 'FeatureCollection', features: [{
...
+- MultiLineString + MultiPoint => no action taken
+
+For the `direct_select` and drawing modes, this cancels the current process.
```
Comment (mcwhittemore) [inline API.md]: What does `cancels the current process` mean? Exit the mode?

File: API.md [hunk 23]
```diff
@@ -244,16 +244,48 @@ var ids = Draw.set({type: 'FeatureCollection', features: [{
...
+
+---
+
+### `.uncombineFeatures() -> Draw`
+
+This invokes the current modes uncombineFeatures event. For the `simple_select` mode this takes the currently features, and for each multi-feature selected, it will split it into its component feature parts. For example:
```
Comment (mcwhittemore) [inline API.md]: `takes the currently features` missing `selected`

File: API.md [hunk 24]
```diff
@@ -244,16 +244,48 @@ var ids = Draw.set({type: 'FeatureCollection', features: [{
...
+
+- MultiLineString (of two parts) => LineString, LineString 
+- MultiLineString (of three parts) => LineString, LineString, LineString
+- MultiLineString (of two parts), Point => LineString, LineString, Point
+- LineString => LineString
+
+For the `direct_select` and drawing modes, this cancels the current process.
```
Comment (mcwhittemore) [inline API.md]: Like `combine`, I'm not sure what this means.

File: src/options.js [hunk 25]
```diff
@@ -16,14 +16,18 @@ const showControls = {
   point: true,
   line_string: true,
   polygon: true,
-  trash: true
+  trash: true,
+  merge_features: true,
```
Comment (mcwhittemore) [inline src/options.js]: this should change to combine

File: src/options.js [hunk 26]
```diff
@@ -16,14 +16,18 @@ const showControls = {
...
+  split_features: true
 };
 
 const hideControls = {
   point: false,
   line_string: false,
   polygon: false,
-  trash: false
+  trash: false,
+  merge_features: false,
```
Comment (mcwhittemore) [inline src/options.js]: This should change to combine

File: test/api.test.js [hunk 27]
```diff
@@ -422,6 +422,138 @@ test('Draw.modes', t => {
...
 
+test('Draw.combineFeatures -- multiploygon', t => {
+  const [polygonId] = Draw.add(getGeoJSON('polygon'));
+  const [polygon2Id] = Draw.add(getGeoJSON('polygon2'));
+  Draw.changeMode('simple_select', { featureIds: [polygonId, polygon2Id]});
+
+  Draw.combineFeatures();
+  t.equals(Draw.getAll().features.length, 1, 'can combine two features');
+  t.equals(Draw.getAll().features[0].geometry.type, 'MultiPolygon', 'can merge two polygons into MultiPolygon');
+  Draw.deleteAll();
```
Comment (mcwhittemore) [inline test/api.test.js]: Lets add a test like

`t.deepEquals(Draw.getAll().features[0].geometry.coordinates[0], geoGeoJSON('polygon').geometry.coordinates)`

So we know that the combine function is working as expected.

Comment (mcwhittemore) [inline test/api.test.js]: We should do this for all the geometry tests that perform a combine

File: test/api.test.js (hunk 14 above)
Comment (mcwhittemore) [inline test/api.test.js]: Oh, sorry, just seeing this. Lets tests that the coordinates from the original features match those in the multifeature

File: src/ui.js [hunk 28]
```diff
@@ -153,6 +154,28 @@ module.exports = function(ctx) {
         }
       });
     }
+
+    if (controls.merge_features) {
+      buttonElements.combine = createControlButton('combineFeatures', {
+        container: controlGroup,
+        className: Constants.classes.CONTROL_BUTTON_COMBINE_FEATURES,
+        title: 'Merge',
```
Comment (mcwhittemore) [inline src/ui.js]: This should change to combine

File: src/ui.js [hunk 29]
```diff
@@ -153,6 +154,28 @@ module.exports = function(ctx) {
...
+          ctx.events.combineFeatures();
+        }
+      });
+    }
+
+    if (controls.split_features) {
+      buttonElements.uncombine = createControlButton('uncombineFeatures', {
+        container: controlGroup,
+        className: Constants.classes.CONTROL_BUTTON_UNCOMBINE_FEATURES,
+        title: 'Split',
```
Comment (mcwhittemore) [inline src/ui.js]: This should change to uncombine

**Comment by mcwhittemore** [comment]
Just played with this and its great! Thanks. Later we should think about https://github.com/mapbox/mapbox-gl-draw/issues/506, but its not something to worry about now. Once the above requests are in, I think this is ready to 📦 and 🚢

**Comment by ghoshkaj** [comment]
@mcwhittemore for review

File: test/mode_handler.test.js [hunk 30]
```diff
@@ -18,7 +18,9 @@ test('returned API', t => {
   t.equal(typeof mh.mouseout, 'function', 'exposes mouseout');
   t.equal(typeof mh.keydown, 'function', 'exposes keydown');
   t.equal(typeof mh.keyup, 'function', 'exposes keyup');
-  t.equal(Object.keys(mh).length, 11, 'no unexpected properties');
+  t.equal(typeof mh.click, 'function', 'exposes combineFeatures');
```
Comment (mcwhittemore) [inline test/mode_handler.test.js]: This should be combine and uncombine

Comment (ghoshkaj) [inline test/mode_handler.test.js]: Do you mean as opposed to combineFeatures and uncombineFeatures?

Comment (ghoshkaj) [inline test/mode_handler.test.js]: Oh you mean as opposed to mh.click.

Comment (ghoshkaj) [inline test/mode_handler.test.js]: Also, should I add back the no unexpected properties test? I don't know why I removed that...

**Comment by ghoshkaj** [comment]
@mcwhittemore hopefully final review? 😄

**Comment by mcwhittemore** [comment]
@ghoshkaj - combine and uncombine work really well, but editing the nodes of multipolygons is broken in this branch.

broken-multi-polygon-editing

Here is the MultiPolygon I'm starting with:

```json
{"type":"FeatureCollection","features":[{"id":"9d2d01c8ff7328bfdc8a8f71f6805254","type":"Feature","properties":{},"geometry":{"coordinates":[[[[112.796493,-22.202828],[115.856788,-34.613611],[123.507525,-34.298186],[131.540799,-31.079466],[140.721684,-37.398149],[149.902569,-38.304289],[154.110475,-25.352866],[144.547053,-17.16226],[143.016905,-11.231591],[139.191537,-17.16226],[135.366168,-14.956769],[136.896315,-11.606555],[131.540799,-11.606555],[128.863041,-15.69463],[126.185283,-13.47352],[120.44723,-19.702389],[112.796493,-22.202828]]],[[[-16.118433,10.818061],[-9.232769,3.994014],[9.894074,4.37553],[14.101980000000196,-16.429830617319084],[18.309886,-35.240873],[26.725697,-33.3448],[42.027172,-12.728366],[39.349413,-3.269173],[53.503278,11.943088],[43.557319,10.442089],[33.993897,27.713653],[33.228823,32.024364],[20.22257,33.312406],[19.840033,30.057992],[10.276611,34.266151],[10.659148,37.670751],[-6.555011,35.831895],[-10.762917,28.724843],[-17.64858,22.520743],[-16.118433,10.818061]]]],"type":"MultiPolygon"}}]}
```

**Comment by mcwhittemore** [comment]
@ghoshkaj - ok, problem fixed. Running tests one last time and then we need to publish this!

**Pull request merged**